### PR TITLE
`app_service` - support preview version `21` for `java_version`

### LINKED ISSUE #24754: Java 21 support for azurerm_linux_function_app
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

Microsoft recently added support for Java 21 runtime in azure functions: https://azure.microsoft.com/en-us/updates/public-preview-azure-functions-support-for-java-21-on-linux/

When can we expect terraform resource to allow for Java 21 as only[ 8, 11 and 17 is currently supported](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_function_app#java_version)?

### New or Affected Resource(s)/Data Source(s)

azurerm_3.90.0

### Potential Terraform Configuration

```hcl
application_stack {
      java_version = "21"
    }
```

### References

_No response_

### LINKED ISSUE #25490: Support for Java 21 for azurerm_linux_web_app
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Description

As evidenced from the screenshot below taken in Azure Portal on 2nd April 2024, Azure Web Apps now support Java 21. For this reason, it would be great if Java 21 is supported in the `azurerm_linux_web_app` resource as well. Would it be possible to implement this any time in the foreseeable future? 🙏🏻 

image

Thank you.

### New or Affected Resource(s)/Data Source(s)

azurerm_linux_web_app

### Potential Terraform Configuration

```hcl
application_stack {
      java_version = 21
    }
```

### References

_No response_

### PR DESCRIPTION
**Pull request by computerlove**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## Description
Add Java 21 as a allowed value for webapp services.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [X] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 

## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevent documentation.
- [x] I have successfully run unit tests with my changes locally. 

## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_linux_web_app` - Support Java 21. ([#25490](https://github.com/hashicorp/terraform-provider-azurerm/issues/25490))
* `azurerm_windows_web_app` - Support Java 21. ([#25490](https://github.com/hashicorp/terraform-provider-azurerm/issues/25490))
* `azurerm_linux_function_app` - Support Java 21. ([#24754](https://github.com/hashicorp/terraform-provider-azurerm/issues/24754))
* `azurerm_windows_function_app` - Support Java 21. ([#24754](https://github.com/hashicorp/terraform-provider-azurerm/issues/24754)) 

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change

## Related Issue(s)
fixes #25490
fixes #24754

### CODE REVIEW

File: internal/services/web/app_service_resource_test.go [hunk 1]
```diff
@@ -1294,6 +1294,43 @@ func TestAccAppService_windowsJava11Java(t *testing.T) {
 	})
 }
 
+
+func TestAccAppService_windowsJava17Java(t *testing.T) {
```
Comment (mbfrahry) [inline internal/services/web/app_service_resource_test.go:1298]: The new value you are trying to add wasn't added to this resource so these new tests are failing. I believe you want to add these tests to `windows_app_service_resource_test.go` or to add the new value to this resources schema

**Review by mbfrahry — DISMISSED** [review]
Hey @computerlove! Thanks for this PR but we've got some failing tests because the new values wasn't added to the correct schema or the tests aren't testing the right resource. 

```
=== RUN   TestAccAppService_windowsJava21Java
=== PAUSE TestAccAppService_windowsJava21Java
=== CONT  TestAccAppService_windowsJava21Java
    testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
        Error: expected site_config.0.java_version to be one of ["1.7" "1.8" "11"], got 21
          with azurerm_app_service.test,
          on terraform_plugin_test.tf line 48, in resource "azurerm_app_service" "test":
          48:     java_version           = "21"
--- FAIL: TestAccAppService_windowsJava21Java (11.10s)
```

```
=== CONT  TestAccAppService_windowsJava17Java
    testcase.go:113: Step 1/2 error: Error running pre-apply refresh: exit status 1
        Error: expected site_config.0.java_version to be one of ["1.7" "1.8" "11"], got 17
          with azurerm_app_service.test,
          on terraform_plugin_test.tf line 48, in resource "azurerm_app_service" "test":
          48:     java_version           = "17"
--- FAIL: TestAccAppService_windowsJava17Java (11.67s)
```

File: internal/services/web/app_service_resource_test.go (hunk 1 above)
Comment (computerlove) [reply internal/services/web/app_service_resource_test.go]: Sorry about that. The change attempt was done by searching for "java_version" and not paying to much attention to what was changes. I was going to try to run the tests, but could not find any description of what ARM_CLIENT_ID and ARM_CLIENT_SECRET, and gave up.

I have redone the changes now, following the pointers from (windows|linux)_web_app_resource.go and (windows|linux)_function_app_resource.go

**Comment by computerlove** [comment]
@mbfrahry When I update the conflict in CHANGELOG.md, should the change be amended and force pushed, or is a merge commit OK?

**Comment by stephybun** [comment]
@computerlove, we take care of updating the CHANGELOG.md after a PR has been merged, so if there are any changes in there they should be removed. We recommend rebasing your changes on top of main, but merging changes from main into your branch is also acceptable.

**Comment by computerlove** [comment]
Ok, I have rebased and pushed with not changes to CHANGELOG.md.

**Review by stephybun — COMMENTED** [review]
Thanks @computerlove! I think we're also missing a test for linux function app?

**Comment by computerlove** [comment]
Yeah, looks like it. 
Added one now.

File: internal/services/appservice/linux_web_app_resource_test.go [hunk 2]
```diff
@@ -1185,6 +1185,38 @@ func TestAccLinuxWebApp_withJava8JBOSSEAP73(t *testing.T) {
...
+func TestAccLinuxWebApp_withJre21Java(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_linux_web_app", "test")
+	r := LinuxWebAppResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.java(data, "21", "JAVA", "21"),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key("site_config.0.linux_fx_version").HasValue("JAVA|21-java21"),
```
Comment (stephybun) [inline internal/services/appservice/linux_web_app_resource_test.go:1213]: This check fails:
```
------- Stdout: -------
=== RUN   TestAccLinuxWebApp_withJre21Java
=== PAUSE TestAccLinuxWebApp_withJre21Java
=== CONT  TestAccLinuxWebApp_withJre21Java
    testcase.go:130: Step 1/3 error: Check failed: Check 2/2 error: azurerm_linux_web_app.test: Attribute 'site_config.0.linux_fx_version' expected "JAVA|21-java21", got "JAVA|21-21"
--- FAIL: TestAccLinuxWebApp_withJre21Java (143.39s)
FAIL
```

**Review by stephybun — CHANGES_REQUESTED** [review]
Most of the tests look good, just one is failing, the Linux Web App test needs to have the check updated

File: internal/services/appservice/linux_web_app_resource_test.go (hunk 2 above)
Comment (computerlove) [reply internal/services/appservice/linux_web_app_resource_test.go]: Updated now.

**Review by katbyte — CHANGES_REQUESTED** [review]
looks like we have a test failure here
```
------- Stdout: -------
=== RUN   TestAccLinuxWebApp_withJre21Java
=== PAUSE TestAccLinuxWebApp_withJre21Java
=== CONT  TestAccLinuxWebApp_withJre21Java
    testcase.go:130: Step 1/3 error: After applying this test step, the non-refresh plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # azurerm_linux_web_app.test will be updated in-place
          ~ resource "azurerm_linux_web_app" "test" {
                id                                             = "/subscriptions/*******/resourceGroups/acctestRG-240815145037738884/providers/Microsoft.Web/sites/acctestWA-240815145037738884"
                name                                           = "acctestWA-240815145037738884"
                # (20 unchanged attributes hidden)
        
              ~ site_config {
                    # (24 unchanged attributes hidden)
        
                  ~ application_stack {
                      + java_version        = "21"
                        # (2 unchanged attributes hidden)
                    }
                }
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccLinuxWebApp_withJre21Java (250.56s)
FAIL
```

**Comment by computerlove** [comment]
Yeah, no idea what that means. 
Changing the expected version string fails in the expected way. So it seems that this is something else.

**Comment by groupcard-ruben-timmermans** [comment]
any update?

**Comment by computerlove** [comment]
No. 
Someone have to find out what `After applying this test step, the non-refresh plan was not empty.` means and why it appears on the new test for java 21, and not for the identical test for java 17.

**Comment by StefanSchoof** [comment]
My interpretation is that, the test is apply the changes and then makes a [terraform plan -refresh=false](https://developer.hashicorp.com/terraform/cli/commands/plan#planning-options) and expects that the plan need no changes. (But why this happens only for java 21, I have no clue)

File: internal/services/appservice/linux_web_app_resource_test.go [hunk 3]
```diff
@@ -1185,6 +1185,38 @@ func TestAccLinuxWebApp_withJava8JBOSSEAP73(t *testing.T) {
...
+func TestAccLinuxWebApp_withJre21Java(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_linux_web_app", "test")
+	r := LinuxWebAppResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.java(data, "21", "JAVA", "21"),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+				check.That(data.ResourceName).Key("site_config.0.linux_fx_version").HasValue("JAVA|21-21"),
```
Comment (erboucherie) [inline internal/services/appservice/linux_web_app_resource_test.go:1213]: Don't you have a typo here ? Should be `JAVA|21-java21`

Comment (computerlove) [reply internal/services/appservice/linux_web_app_resource_test.go:1213]: Changing to `JAVA|21-java21` results in 
> testcase.go:173: Step 1/3 error: Check failed: Check 2/2 error: azurerm_linux_web_app.test: Attribute 'site_config.0.linux_fx_version' expected "JAVA|21-java21", got "JAVA|21-21"

Comment (jackofallops) [reply internal/services/appservice/linux_web_app_resource_test.go:1213]: This is actually due to some missing code to account for this change. Can you extend the `decodeApplicationStackLinux()` function to accommodate the new version `21`
[here](https://github.com/hashicorp/terraform-provider-azurerm/blob/c8a5ffba4cbc370d5c89160cbdb4469c89d76b83/internal/services/appservice/helpers/fx_strings.go#L54)
and similarly `JavaLinuxFXStringBuilder()` [here](https://github.com/hashicorp/terraform-provider-azurerm/blob/c8a5ffba4cbc370d5c89160cbdb4469c89d76b83/internal/services/appservice/helpers/fx_strings.go#L265)

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @computerlove - Thanks for this PR. Looking over there's some additional changes needed to accommodate the new version, which I've noted for you below.  If you can take a look it should get the test passing, and the fx_version should then be correctly set to `JAVA|21-java21` and the tests should pass.

**Comment by computerlove** [comment]
Thank you! Will take a look in a few days.

**Comment by computerlove** [comment]
```
marlil@marlil:~/src/terraform-provider-azurerm$ make acctests SERVICE='appservice' TESTARGS='-run=TestAccLinuxWebApp_withJre21Java' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/appservice -run=TestAccLinuxWebApp_withJre21Java -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccLinuxWebApp_withJre21Java
=== PAUSE TestAccLinuxWebApp_withJre21Java
=== CONT  TestAccLinuxWebApp_withJre21Java
--- PASS: TestAccLinuxWebApp_withJre21Java (251.43s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    251.461s
marlil@marlil:~/src/terraform-provider-azurerm$ make acctests SERVICE='appservice' TESTARGS='-run=TestAccLinuxFunctionApp_appStackJava21' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/appservice -run=TestAccLinuxFunctionApp_appStackJava21 -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccLinuxFunctionApp_appStackJava21
=== PAUSE TestAccLinuxFunctionApp_appStackJava21
=== CONT  TestAccLinuxFunctionApp_appStackJava21
--- PASS: TestAccLinuxFunctionApp_appStackJava21 (313.70s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    313.723s
marlil@marlil:~/src/terraform-provider-azurerm$ make acctests SERVICE='appservice' TESTARGS='-run=TestAccLinuxWebApp_withJre17Java' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/appservice -run=TestAccLinuxWebApp_withJre17Java -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccLinuxWebApp_withJre17Java
=== PAUSE TestAccLinuxWebApp_withJre17Java
=== CONT  TestAccLinuxWebApp_withJre17Java
--- PASS: TestAccLinuxWebApp_withJre17Java (251.59s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    251.611s
marlil@marlil:~/src/terraform-provider-azurerm$ make acctests SERVICE='appservice' TESTARGS='-run=TestAccWindowsFunctionApp_appStackJava17' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/appservice -run=TestAccWindowsFunctionApp_appStackJava17 -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccWindowsFunctionApp_appStackJava17
=== PAUSE TestAccWindowsFunctionApp_appStackJava17
=== CONT  TestAccWindowsFunctionApp_appStackJava17
--- PASS: TestAccWindowsFunctionApp_appStackJava17 (309.14s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    309.160s
marlil@marlil:~/src/terraform-provider-azurerm$ make acctests SERVICE='appservice' TESTARGS='-run=TestAccWindowsFunctionApp_appStackJava21' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/appservice -run=TestAccWindowsFunctionApp_appStackJava21 -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccWindowsFunctionApp_appStackJava21
=== PAUSE TestAccWindowsFunctionApp_appStackJava21
=== CONT  TestAccWindowsFunctionApp_appStackJava21
--- PASS: TestAccWindowsFunctionApp_appStackJava21 (318.06s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    318.085s
marlil@marlil:~/src/terraform-provider-azurerm$ make acctests SERVICE='appservice' TESTARGS='-run=TestAccWindowsWebApp_withJava21Embedded' TESTTIMEOUT='60m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/appservice -run=TestAccWindowsWebApp_withJava21Embedded -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccWindowsWebApp_withJava21Embedded
=== PAUSE TestAccWindowsWebApp_withJava21Embedded
=== CONT  TestAccWindowsWebApp_withJava21Embedded
--- PASS: TestAccWindowsWebApp_withJava21Embedded (253.43s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/appservice    253.451s

```

**Review by erboucherie — APPROVED** [review]
Sounds good to me .

**Review by jackofallops — APPROVED** [review]
Thanks for the updates @computerlove - This LGTM now 👍

**Comment by jackofallops** [comment]
GHA Unit Test failure due to unrelated test code on merge comparison branch that has been fixed in `main`.
image

**Pull request merged**